server/statistics/topn.go

```diff
@@ -24,27 +24,30 @@ import (
 type TopNItem interface {
 	// ID is used to check identity.
 	ID() uint64
-	// Less tests whether the current item is less than the given argument.
-	Less(then TopNItem) bool
+	// Less tests whether the current item is less than the given argument in the `k`th dimension.
+	Less(k int, than TopNItem) bool
 }
 
-// TopN maintains the N largest items.
+// TopN maintains the N largest items of multiple dimensions.
 type TopN struct {
 	rw     sync.RWMutex
 	n      int
-	topn   *indexedHeap
-	rest   *indexedHeap
+	topns  []*singleTopN
 	ttlLst *ttlList
 }
 
-// NewTopN returns a TopN with given TTL.
-func NewTopN(n int, ttl time.Duration) *TopN {
-	return &TopN{
+// NewTopN returns a k-dimensional TopN with given TTL.
+func NewTopN(k, n int, ttl time.Duration) *TopN {
+	ret := &TopN{
 		n:      maxInt(n, 1),
-		topn:   newTopNHeap(n),
-		rest:   newRevTopNHeap(n),
+		topns:  make([]*singleTopN, k),
 		ttlLst: newTTLList(ttl),
 	}
+	k = maxInt(k, 1)
```

Comment:
Why do we need it?

Reply:
To make sure `k >= 1` after this line.

Reply:
But if `k == 0`, it will cause panic?

Reply:
I see, it is bug.

Reply:
Fixed. I decided to make it panic if `k <= 0 || n <= 0`.